lib/friends/event.rb

```diff
@@ -132,9 +133,21 @@ def update_location_name(old_name:, new_name:)
     end
 
     # @param location [Location] the location to test
-    # @return [Boolean] true iff this activity includes the given location
+    # @return [Boolean] true if activity has location in description or it equals implicit location
     def includes_location?(location)
-      @description.scan(/(?<=_)[^_]+(?=_)/).include? location.name
+      location_in_description?(location) || location_is_implicit?(location)
+    end
+
+    def location_in_description?(location)
+      (@description.scan(/(?<=_)[^_]+(?=_)/).include? location.name)
```

Comment:
Since this is so close to the behavior of `location_names`, how would you feel about just using that here? e.g. `location_names.include? location.name`.

Reply:
Also if we do that I think it makes sense to memoize the result of `location_names`, e.g.

```
def location_names
  @location_names ||= @description.scan(/(?<=_)\w[^_]*(?=_)/).uniq
end
```

since this is always called in `set_implicit_locations` and then also called again when filtering (e.g. `friends list activities --in Paris`) _or_ writing the file (the `clean` method in Introvert). But please check my thinking on that!

Reply:
> Since this is so close to the behavior of `location_names`, how would you feel about just using that here? e.g. `location_names.include? location.name`.

Done!

> Also if we do that I think it makes sense to memoize the result of `location_names`, e.g.
> 
> ```
> def location_names
>   @location_names ||= @description.scan(/(?<=_)\w[^_]*(?=_)/).uniq
> end
> ```
> 
> since this is always called in `set_implicit_locations` and then also called again when filtering (e.g. `friends list activities --in Paris`) _or_ writing the file (the `clean` method in Introvert). But please check my thinking on that!

Some tests are failing when I memoize the result of `location_names`, namely the test to do with renaming locations.
Whenever the `rename location` command is called, `the set_implicit_location` command is run first and the `clean` command runs at the end. Both of these methods call `location_names`.
When we memoize the result of `location_names` to `@location_names`, and run the `rename location` command, `set_implicit_location` runs `location_name` and sets `@location` to whatever the current location (eg London) in the description is. Then the current location in the description and `### Locations` section is renamed (to eg Paris). Then the `clean` command runs `location_names` command again to check for any disparities between locations in activities and locations in `### Locations` - and because `@locations` is set to London, the clean command will think there is a disparity and add London as a new location.

I'm going to therefore leave this as is, unless you have another suggestion?

Reply:
Ah, interesting! Thanks for digging into that. I think you're right that leaving it as-is is a good way to proceed. (And wow am I glad this codebase has so many good tests!)

Reply:
Awesome - I'm glad there are loads of tests too, makes my task so much easier 😄